controllers/agentaction_controller.go

```diff
@@ -713,14 +711,27 @@ func (r *AgentActionReconciler) getAgentVolumes(action *porterv1.AgentAction, pv
 		},
 		)
 	}
-
-	for _, volume := range action.Spec.Volumes {
-		volumes = append(volumes, volume)
+	if !action.IsAgentConfig() {
+		claimName := agentCfg.GetPVCName(action.Namespace)
```

Comment:
What if the AgentConfig has a non Succeeded status? Does an AgentAction that is using an AgentConfig that is still being created block until that AgentConfig job finishes? If an AgentAction is using an AgentConfig that failed to install the plugins will that AgentAction fail as well?

Reply:
Yes, the AgentAction is blocked until the AgentConfig finishes installing plugins and the plugin volume is ready.
If an AgentConfig fails, the plugin volume claim would not exist, I would assume the following agent action would timeout eventuall(I tried to google information on the timeout but I haven't able to confirm the timeout behavior).
Do you have any suggestion on how to we can better handle the failure case here? To me, if the plugins are not ready, there's probably some serious issue going on and the operator should not try to continue working.

Reply:
I think the action should fail if the AgentConfig associated with that action didn't successfully install the plugins. Just wanted to make sure that was how this worked I didn't see how that was happening in the code currently. We could also check the associated status of the AgentConfig resource we get with the AgentAction and if its in "Failed" state then fail the AgentAction. I'm not sure on the k8s internal behavior though if you try to bind a PVC that doesn't exist to a job will it wait for that to exist before attempting the job or will it fail immediately?

This is the behavior I would expect:
* Given: AgentConfig is still running
   * When: An AgentAction is created that uses that AgentConfig
      * Then: The AgentAction will wait until the AgentConfig reaches a finished state before creating the porter job for that AgentAction
* Given: AgentConfig resource is running
   * When: 
      * Then: Any waiting AgentActions relying on that AgentConfig will fail
      * And: Any new AgentActions that reference that AgentConfig will fail
* Given: No AgentConfig
   * When: A new AgentAction is created
      * Then: A default AgentConfig is created
      * And: The AgentAction waits for that AgentConfig to reach a completed state

Do those sound reasonable and is this currently what the behavior is?

Reply:
Thanks for the clarification. In your list of the expected behaviors, do you mind to explain what's the difference between the first and the second scenarios? 
Also I'm not sure an AgentAction has an associated AgentConfig once all the configs(AgentConfig and porter configs) are resolved. On this line https://github.com/getporter/operator/blob/5ac5ff17a3473c663eb1f24cc19305b75858188c/controllers/agentaction_controller.go#L468, the `resolveAgentConfig` function merges default values with the AgentConfig defined on the AgentAction. The plugins could be installed by a namespace level AgentConfig not the one defined on the AgentAction.
To answer your question above, k8s will wait to attempt a job if a binded PVC doesn't exit yet and will proceed once the PVC is ready.

Reply:
Whoops sorry that was supposed to be "When: The AgentConfig AgentAction fails".

So there's no way for an AgentAction to resolve to a specific AgentConfig based on what is specified in its AgentConfig, that's just a reference to the AgentConfigSpec not an actual resource correct? That makes sense so it would be tough to know if that corresponding AgentAction succeeded. I just don't want to get into a point where an Installation can hang either indefinitely or for a very long time waiting for a PVC that will never appear

Reply:
Hey, I have added a new status field on AgentConfig resource to indicate whether an AgentConfig is ready(meaning the plugins has been successfully installed). Any agent action job will only be created once the AgentConfig is ready. If it's not ready, the agent action reconcile loop will do the retry